zk/stages/stage_sequence_execute_utils.go

```diff
@@ -0,0 +1,324 @@
+
+	transactionGasLimit = 30000000
+
+	yieldSize = 100 // arbitrary number defining how many transactions to yield from the pool at once
+)
+
+var (
+	noop            = state.NewNoopWriter()
+	blockDifficulty = new(big.Int).SetUint64(0)
+)
+
+type HasChangeSetWriter interface {
+	ChangeSetWriter() *state.ChangeSetWriter
+}
+
+type ChangeSetHook func(blockNum uint64, wr *state.ChangeSetWriter)
+
+type SequenceBlockCfg struct {
+	db            kv.RwDB
+	batchSize     datasize.ByteSize
+	prune         prune.Mode
+	changeSetHook ChangeSetHook
+	chainConfig   *chain.Config
+	engine        consensus.Engine
+	zkVmConfig    *vm.ZkConfig
+	badBlockHalt  bool
+	stateStream   bool
+	accumulator   *shards.Accumulator
+	blockReader   services.FullBlockReader
+
+	dirs      datadir.Dirs
+	historyV3 bool
+	syncCfg   ethconfig.Sync
+	genesis   *types.Genesis
+	agg       *libstate.AggregatorV3
+	zk        *ethconfig.Zk
+
+	txPool   *txpool.TxPool
+	txPoolDb kv.RwDB
+}
+
+func StageSequenceBlocksCfg(
+	db kv.RwDB,
+	pm prune.Mode,
+	batchSize datasize.ByteSize,
+	changeSetHook ChangeSetHook,
+	chainConfig *chain.Config,
+	engine consensus.Engine,
+	vmConfig *vm.ZkConfig,
+	accumulator *shards.Accumulator,
+	stateStream bool,
+	badBlockHalt bool,
+
+	historyV3 bool,
+	dirs datadir.Dirs,
+	blockReader services.FullBlockReader,
+	genesis *types.Genesis,
+	syncCfg ethconfig.Sync,
+	agg *libstate.AggregatorV3,
+	zk *ethconfig.Zk,
+
+	txPool *txpool.TxPool,
+	txPoolDb kv.RwDB,
+) SequenceBlockCfg {
+	return SequenceBlockCfg{
+		db:            db,
+		prune:         pm,
+		batchSize:     batchSize,
+		changeSetHook: changeSetHook,
+		chainConfig:   chainConfig,
+		engine:        engine,
+		zkVmConfig:    vmConfig,
+		dirs:          dirs,
+		accumulator:   accumulator,
+		stateStream:   stateStream,
+		badBlockHalt:  badBlockHalt,
+		blockReader:   blockReader,
+		genesis:       genesis,
+		historyV3:     historyV3,
+		syncCfg:       syncCfg,
+		agg:           agg,
+		zk:            zk,
+		txPool:        txPool,
+		txPoolDb:      txPoolDb,
+	}
+}
+
+type stageDb struct {
+	tx          kv.RwTx
+	hermezDb    *hermez_db.HermezDb
+	eridb       *db2.EriDb
+	stateReader *state.PlainStateReader
+	smt         *smtNs.SMT
+}
+
+func newStageDb(tx kv.RwTx) *stageDb {
+	sdb := &stageDb{
+		tx:          tx,
+		hermezDb:    hermez_db.NewHermezDb(tx),
+		eridb:       db2.NewEriDb(tx),
+		stateReader: state.NewPlainStateReader(tx),
+		smt:         nil,
+	}
+	sdb.smt = smtNs.NewSMT(sdb.eridb)
+
+	return sdb
+}
+
+func prepareForkId(cfg SequenceBlockCfg, lastBatch, executionAt uint64, hermezDb *hermez_db.HermezDb) (uint64, error) {
+	var forkId uint64 = 0
+	var err error
+
+	if executionAt == 0 {
+		// capture the initial sequencer fork id for the first batch
+		forkId = cfg.zk.SequencerInitialForkId
+		if err := hermezDb.WriteForkId(1, forkId); err != nil {
+			return forkId, err
+		}
+		if err := hermezDb.WriteForkIdBlockOnce(uint64(forkId), 1); err != nil {
+			return forkId, err
+		}
```

Comment:
Just double checking here, didn't we at one point have a loop to populate all of the historical fork ids as well at block 1?  Some of the logic checks "if forkid == 7" but if we start the chain at fork id 8 that fork is never written so the logic fails.  Might be handled elsewhere now, just double checking.

Reply:
Value has fixed that globally so we should now be safe